### PR TITLE
Specify `Warning` header usage

### PR DESCRIPTION
**Pull request by imjasonh**
Fixes https://github.com/opencontainers/distribution-spec/issues/390

This is _heavily_ inspired by Kubernetes' support for warnings returned by the API server and surfaced by clients like `kubectl`: https://kubernetes.io/blog/2020/09/03/warnings/

tl;dr: Registries MAY return informational warning messages in the `Warning` header, as described in https://www.rfc-editor.org/rfc/rfc7234#section-5.5, with the added restriction that the `warn-code` must be 299 and the `warn-agent` must be unset (`-`). Clients SHOULD surface those messages to the user in an unobtrusive and helpful way -- how they do that is up to them.

Blatantly cribbing further helpful advice from https://github.com/opencontainers/distribution-spec/issues/390#issuecomment-1472827380:

> From https://kubernetes.io/blog/2020/09/03/warnings/#admission-webhooks
>
> > If you are implementing a webhook that returns a warning message, here are some tips:
> >
> > - Don't include a "Warning:" prefix in the message (that is added by clients on output)
> > - Use warning messages to describe problems the client making the API request should correct or be aware of
> > - Be brief; limit warnings to 120 characters if possible

and from https://github.com/opencontainers/distribution-spec/issues/390#issuecomment-1472832235

>  One more side note: a key provision of the warnings RFC is that for warnings returned with code 299: "A system receiving this warning MUST NOT take any automated action." Kubernetes leans heavily on this to assume it is always safe to return a warning or stop returning a warning without backwards compatibility implications. You would think it goes without saying, but warning header contents are not APIs.

cc @BenTheElder @liggitt 

@jonjohnsonjr @jdolitsky @sudo-bmitch @sajayantony

### CODE REVIEW

File: spec.md [hunk 1]
```diff
@@ -758,6 +759,24 @@ The `code` field MUST be one of the following:
 | code-13 | `UNSUPPORTED`           | the operation is unsupported                               |
 | code-14 | `TOOMANYREQUESTS`       | too many requests                                          |
 
+#### Warnings
+
+Registry implementations MAY include informational warnings in `Warning` headers, as described in https://www.rfc-editor.org/rfc/rfc7234#section-5.5.
+
+If included, `Warning` headers MUST specify a `warn-code` of `299` and a `warn-agent` of `-`, and MUST NOT specify a `warn-date` value.
+
+A registry MUST NOT send more than 4096 bytes of warning data.
```
Comment (sudo-bmitch) [inline spec.md:768]: Is this limit per header, or the sum of all headers in a response?

Comment (liggitt) [reply spec.md:768]: for kubernetes, it was the sum... we were paranoid about various proxies or clients setting small limits on how much header content they would accept and breaking responses. The most conservative defaults I found ranged from 8kb-32kb, so I halved the most conservative default I found and capped warning headers to that.

File: spec.md [hunk 2]
```diff
@@ -758,6 +759,24 @@ The `code` field MUST be one of the following:
...
+
+Example warning headers:
+
+```
+Warning: 299 - "Your auth token will expire in 30 seconds."
+Warning: 299 - "This registry endpoint is deprecated and will be removed soon."
+Warning: 299 - "This image is deprecated and will be removed soon."
+```
+
+If a client receives `Warning` response headers, it SHOULD report the warnings to the user in an unobtrusive way.
```
Comment (BenTheElder) [inline spec.md:778]: Is there somewhere we can document suggestions about deduping etc?

Should we add an explicit note that these are informational only and clients should not interpret them?

File: spec.md [hunk 3]
```diff
@@ -758,6 +759,24 @@ The `code` field MUST be one of the following:
 | code-13 | `UNSUPPORTED`           | the operation is unsupported                               |
 | code-14 | `TOOMANYREQUESTS`       | too many requests                                          |
 
+#### Warnings
+
+Registry implementations MAY include informational warnings in `Warning` headers, as described in https://www.rfc-editor.org/rfc/rfc7234#section-5.5.
```
Comment (sudo-bmitch) [inline spec.md:764]:
```suggestion
Registry implementations MAY include informational warnings in `Warning` headers, as described in [RFC 7234](https://www.rfc-editor.org/rfc/rfc7234#section-5.5).
```

File: spec.md (hunk 1 above)
Comment (sudo-bmitch) [reply spec.md:768]:
```suggestion
A registry MUST NOT send more than 4096 bytes of warning data from all headers combine.
```

File: spec.md (hunk 2 above)
Comment (sudo-bmitch) [reply spec.md:778]:
```suggestion
If a client receives `Warning` response headers, it SHOULD report the warnings to the user in an unobtrusive way.
Clients SHOULD deduplicate warnings from multiple associated responses.
```

File: spec.md [hunk 4]
```diff
@@ -758,6 +759,25 @@ The `code` field MUST be one of the following:
 | code-13 | `UNSUPPORTED`           | the operation is unsupported                               |
 | code-14 | `TOOMANYREQUESTS`       | too many requests                                          |
 
+#### Warnings
+
+Registry implementations MAY include informational warnings in `Warning` headers, as described in [RFC 7234](https://www.rfc-editor.org/rfc/rfc7234#section-5.5).
+
+If included, `Warning` headers MUST specify a `warn-code` of `299` and a `warn-agent` of `-`, and MUST NOT specify a `warn-date` value.
+
+A registry MUST NOT send more than 4096 bytes of warning data.
```
Comment (imjasonh) [inline spec.md:768]:
```suggestion
A registry MUST NOT send more than 4096 bytes of warning data from all headers combined.
```

File: spec.md [hunk 5]
```diff
@@ -758,6 +759,25 @@ The `code` field MUST be one of the following:
...
+
+```
+Warning: 299 - "Your auth token will expire in 30 seconds."
+Warning: 299 - "This registry endpoint is deprecated and will be removed soon."
+Warning: 299 - "This image is deprecated and will be removed soon."
+```
+
+If a client receives `Warning` response headers, it SHOULD report the warnings to the user in an unobtrusive way.
+Clients SHOULD deduplicate warnings from multiple associated responses.
+
```
Comment (BenTheElder) [inline spec.md:780]: "Clients MUST NOT depend on warning contents."?

File: spec.md [hunk 6]
```diff
@@ -23,6 +23,7 @@
 - [API](#api)
 	- [Endpoints](#endpoints)
 	- [Error Codes](#error-codes)
+  - [Warnings](#warnings)
```
Comment (jdolitsky) [inline spec.md:26]: something off with the spacing here

File: spec.md [hunk 7]
```diff
@@ -23,6 +23,7 @@
 - [API](#api)
 	- [Endpoints](#endpoints)
 	- [Error Codes](#error-codes)
+    - [Warnings](#warnings)
```
Comment (jdolitsky) [inline spec.md:26]: still seems off compared to Error Codes?

Comment (sudo-bmitch) [reply spec.md:26]: Did we mix tabs and spaces in the TOC?

**Review by jdolitsky — APPROVED** [review]

**Review by sudo-bmitch — APPROVED** [review]
LGTM

**Review by jonjohnsonjr — APPROVED** [review]

**Review by vbatts — APPROVED** [review]

**Pull request merged**